ios-target/View/Home/TargetFormViewController.swift

```diff
@@ -10,7 +10,65 @@ import UIKit
 
 class TargetFormViewController: ModalViewController {
 
+  @IBOutlet weak var areaTextField: UITextFieldPadding!
+  @IBOutlet weak var titleTextField: UITextFieldPadding!
+  @IBOutlet weak var topicTextField: UITextFieldPadding!
+  @IBOutlet weak var createTargetButton: UIButton!
+  
+  var viewModel = TargetFormViewModel()
+  
+  weak var delegate: TargetActionsDelegate?
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    viewModel.delegate = self
+    setCreateTargetButton(enabled: false)
+  }
+  
+  @IBAction func formEditingChange(_ sender: UITextField) {
+    let newValue = sender.text ?? ""
+    switch sender {
+    case areaTextField:
+      viewModel.targetArea = Double(newValue) ?? 0
+    case titleTextField:
+      viewModel.targetTitle = newValue
+    case topicTextField:
+      viewModel.targetTopic = Int(newValue) ?? 0
+    default: break
+    }
+  }
+  
   @IBAction func tapSaveTargetButton(_ sender: Any) {
+    viewModel.createTarget()
+  }
+  
+  func setCreateTargetButton(enabled: Bool) {
+    createTargetButton.alpha = enabled ? 1 : 0.5
+    createTargetButton.isEnabled = enabled
+  }
+}
+
+extension TargetFormViewController: TargetFormViewModelDelegate, TargetActionsDelegate {
+  func newTargetCreated(targets: [Target]) {
+    delegate?.newTargetCreated(targets: targets)
     dismiss(animated: true)
   }
+  
+  func didUpdateState() {
+    switch viewModel.state {
+    case .idle:
+      UIApplication.hideNetworkActivity()
+    case .loading:
+      UIApplication.showNetworkActivity()
+    case .error(let errorDescription):
+      UIApplication.hideNetworkActivity()
+      showMessage(title: "Error", message: errorDescription)
+    case .saved:
+      UIApplication.hideNetworkActivity()
+    }
```

Comment:
``` swift

func didUpdateState() {
  if viewModel.state == .loading {
    // show activity 
} else {
  //hide activity
}

if state == .error(description) {
 //show error
}

Reply:
I can't equate state to .error(description) as Swift doesn't synthesize the necessary equality operators automatically. I defined the default to hide activity for .saved and .idle and left .loading and .error un separate cases. What do you think?

Reply:
hmm you're trying `if state == .error(let description)`?  maybe the `let` was missing, should be the same as in the switch, I can't see why the compiler would have an issue with that. But if that's the case, I don't see the point of implementing `==` just for this

Reply:
yes, tried using `if state == .error(let description)` and it received this error: `Binary operator '==' cannot be applied to operands of type 'TargetFormViewModelState' and '_'` and others.

Reply:
I think the syntax is `if case .error(description) = state { ... }`

Reply:
right! 🤦‍♂